crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -785,17 +785,21 @@ where
                     }
                     self.visit_body(&clause.body);
                 }
-                for post_clause_state in post_clauses {
-                    self.flow_merge(post_clause_state);
-                }
                 let has_else = node
                     .elif_else_clauses
                     .last()
                     .is_some_and(|clause| clause.test.is_none());
                 if !has_else {
-                    // if there's no else clause, then it's possible we took none of the branches,
-                    // and the pre_if state can reach here
-                    self.flow_merge(pre_if);
+                    // if there's no else clause, we should act as if we had an empty else clause,
+                    // by again following the same logic as above for elif/else clauses.
+                    post_clauses.push(self.flow_snapshot());
```

Comment:
Duplicated the logic in above `for` loop of `node.elif_else_clauses`. Another approach would be to have a dummy `else` clause. But I wasn't sure if a dummy node with missing range would cause issues down the line.

Reply:
Constructing an actual dummy node will cause issues, and I don't think it's necessary. But I do think we _could_ remove the logic duplication here. Instead of iterating over clauses above, we could iterate over `(test, body)` pairs, where both elements of the pair are `Option`. For a normal `else` clause, `test` would be `None` but `body` would be `Some`; if there's no `else` clause, we'd insert a `(None, None)` at the end.

That said, without trying to write it, I'm not sure if the code necessary to do this might end up being larger / harder to understand than the duplication you currently have here. Open to whatever you think is best, having worked on this.

Reply:
Thank you for the suggestion. This is better. Slight change I did was with `body`. It is a vector/slice. So I opted to using empty slice instead if `Option`. I have updated the code now.